Neural Lace Changes

BurpleBineapple opened the pull request:
## About The Pull Request
This PR changes the way neural laces work, and fixes gibbing not dropping organs.

## Why It's Good For The Game
Originally, this PR was to fix a headgibbing bug, but at Carl's request it has changed to be a rework of laces in general. Laces can now be destroyed if they take enough damage. A lace being destroyed will instantly kill the owner.

Laces are not affected by EMPs or toxins still. However being shot in the head repeatedly will damage your lace. You are more likely to die suddenly from head trauma with a lace than without.

Future PR may need to change lace hit chances or its health. Won't know for sure without some proper gameplay. Consider test merging this before committing to it.

Lace dropping upon headgib:
![image](https://user-images.githubusercontent.com/67706292/159076661-2ce7b6dd-454e-4ae0-9bbf-467b13f4f88e.png)

Lace destruction due to damage:
![image](https://user-images.githubusercontent.com/67706292/159140850-78143f14-5833-406e-84bc-e9de317dd917.png)

## Did You Test It?
Yes.

## Authorship
PurplePineapple#0001

## Changelog

:cl:
tweak: Organs are now guaranteed to drop when you gib a body part with them.
bugfix: Neural laces can now be non-functional if they take enough damage. A neural lace being destroyed kills its owner.
/:cl:

<!--
Please make sure to detail the changes addressed in this PR on your description and on your title.
Jokes are fine, but the Pull Request needs to be easy to locate and read. Be clear and concise!

Here are the tags supported by changelog:
* rscadd - Adding a feature.
* rscdel - Removing a feature.
* tweak - Changing an existing feature.
* bugfix - Fixing an intended functionality that is not working, or correcting an oversight.
* maptweak - Changing something on a map, or adding a new away site. In 99% of cases, all map changes are maptweak.
* spellcheck - Spelling and grammar fixes.
Uncommon tags:
* admin - Adding, removing or changing administrative tools.
* balance - Changing an existing feature in such a way that it may broadly impact game balance; usually reserved for larger changes.
* soundadd - Adding new sounds, usually covered by rscadd unless you're only adding the sounds themselves.
* sounddel - Ditto as above with rscdel
* imageadd - Adding new icons; same situation as soundadd - usually you're adding something that uses these icons, so this isn't needed
* imagedel - Ditto as above.
* experiment - For experimental changes and tests that are intended to be temporary.
* wip - For works in progress. You probably won't get away with using this one.

Here's a changelog example:
:cl: Yourname
rscadd: Adds a new energy weapon, along with sprites and sound effects.
sounddel: Removes the unused X song
imageadd: Adds Skrell pictures to the magazines around the ship/station
/:cl:

############################

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one– and preferably only one –obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than right now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea – let's do more of those!
-->

BurpleBineapple committed:
organs now drop when body parts are gibbed

to avoid having laces and MMIs destroyed when the head is destroyed, all body parts now drop organs when gibbed.

Yurek1 commented:
> ## About The Pull Request
> This PR aims to make organs drop when a body part containing them is gibbed.
> 
> ## Why It's Good For The Game
> To avoid having laces and MMIs destroyed when the head is destroyed, all body parts now drop organs when gibbed. After talking with Rectangles regarding #1575, we decided to take a look at the gibbing process. This PR changes it so that organs are guaranteed to drop from gibbed body parts and should eliminate the need to remove headgibbing.
> 
> ![image](https://user-images.githubusercontent.com/67706292/159076661-2ce7b6dd-454e-4ae0-9bbf-467b13f4f88e.png)
> 
> ## Did You Test It?
> Yes.
> 
> ## Authorship
> PurplePineapple#0001
> 
> ## Changelog
> 🆑 tweak: Organs are now guaranteed to drop when you gib a body part with them. /🆑

My one and only issue with this is that cloning relies on the head existing for some reason. They can't be properly cloned without volunteers or workaround fuckery (attaching a different head). Other than that, I'm perfectly fine with this as an alternative to removing gibbing outright.

N8-Toe commented:
Cloning can be done without a head if you graft a head on, including a monkey head

ProbablyCarl commented:
![image](https://user-images.githubusercontent.com/25065997/159078040-97d7460d-da41-45f5-beec-ee4440197769.png)
It's worth noting that this is the serious bug that is currently in place with laces.

ProbablyCarl commented:
![image](https://user-images.githubusercontent.com/25065997/159078122-4bfbcb57-f205-47f9-a894-dc68d46c987d.png)

BurpleBineapple commented:
> ![image](https://user-images.githubusercontent.com/25065997/159078040-97d7460d-da41-45f5-beec-ee4440197769.png) It's worth noting that this is the serious bug that is currently in place with laces.

https://github.com/BoHBranch/BoH-Bay/blob/b95cdc2790810c43329ef9c3b62ab53dd4382cde/code/modules/boh/organs/disintegratable.dm

This may be why?

ProbablyCarl commented:
Probably.
I just remember it being an important piece for balancing around it. Enough trauma should be enough to permanently remove someone from the round, via lace destruction. Right now it's possible with head gibbing due to oversight and this bug.

BurpleBineapple committed:
Neural Laces can now die.

BurpleBineapple commented:
After some checking around, there was a bypass check on the surgery steps for neural laces. Now, if a lace is heavily damaged, it cannot be transplanted just like any other organ.

We may need to buff lace health later but, for now this should work as wanted.

BurpleBineapple commented:
PR has been renamed and description has bene changed to align with the new additions.

BurpleBineapple commented:
DNM for now, still working on the specifics

BurpleBineapple committed:
Update stack.dm

BurpleBineapple committed:
Revert "Update stack.dm"

This reverts commit 206b2decc0ed9721bd805c3243ecbef8518abb49.

BurpleBineapple committed:
Lace destruction now works

Laces can be destroyed if they take enough damage. If you lose your lace, you instantly die.

BurpleBineapple committed:
Test Value for Lace Hit

BurpleBineapple commented:
Now fully functional.
